### PR TITLE
Fix deploy as is VM start after template deletion

### PR DESCRIPTION
**Pull request by harikrishna-patnala**
### Description

This PR fixes the issue #7970 

Fix is to keep the template deploy as is details even after the template is deleted. VM start need to know whether the vapp property is password or not which are saved in template_deploy_as_is_details table.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

1. Registered a VMware template with "Read VM settings from OVA" enabled
2. Created a VM with it
3. Force delete the template
4. Stop and start the VM => succeeded

### CODE REVIEW

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Review by weizhouapache — APPROVED** [review]
code lgtm

I am fine with this change, although this will lead to some records in the database.
I think this is safest way to do, unless we can 100% figure out what details needs to be kept after template is removed.

We probably need to look into the template details as well
```
            // Remove template details
            templateDetailsDao.removeDetails(template.getId());
```

**Comment by weizhouapache** [comment]
> I think the best way is to copy template details to the ivm_instance_details on deploy

@DaanHoogland 
it may require some code changes (e.g. read property in vm deploy-as-is details instead of template deploy-as-is details )

**Comment by DaanHoogland** [comment]
> > I think the best way is to copy template details to the ivm_instance_details on deploy
> 
> @DaanHoogland it may require some code changes (e.g. read property in vm deploy-as-is details instead of template deploy-as-is details )

yes, but that would solve the DB remnance. I am not demanding we change this, just suggesting what could be a better design choice.

**Comment by harikrishna-patnala** [comment]
template_deploy_as_is_details looks like this
image

The corresponding values in user_vm_deploy_as_is_details looks like this
image

So I dont think we can directly copy the template_deploy_as_is_details to user_vm_deploy_as_is_details as it is.
If we have to do that, we need a logic to differentiate the properties entries of templates and their values in the same table "user_vm_deploy_as_is_details", which I'm not sure if it is possible. May be @nvazquez can help here.

Currently we are parsing everything from user_vm_deploy_as_is_details table and passing them to the resource.
https://github.com/shapeblue/cloudstack/blob/4a0636374931e4e6b562def76acb9f8673a7f22a/engine/storage/src/main/java/org/apache/cloudstack/storage/image/deployasis/DeployAsIsHelperImpl.java#L317-L324

To simplify all these, I thought it will be better to keep the template_deploy_as_is_details even after deletion.
Or else like I mentioned in the issue ticket, we need another column in user_vm_deploy_as_is details table which is "is_password", because thats the only field we are looking in template details table.

cc @DaanHoogland @weizhouapache

File: server/src/main/java/com/cloud/template/HypervisorTemplateAdapter.java [hunk 1]
```diff
@@ -665,9 +665,6 @@ public boolean delete(TemplateProfile profile) {
             // Remove template details
             templateDetailsDao.removeDetails(template.getId());
 
-            // Remove deploy-as-is details (if any)
```
Comment (shwstppr) [inline server/src/main/java/com/cloud/template/HypervisorTemplateAdapter.java:668]: I think we atleast need to add a check if the template is used by existing VMs or not. If it is not used then I don't see a problem removing deployasis details.
And maybe preserve specific details if needed.

Comment (harikrishna-patnala) [reply server/src/main/java/com/cloud/template/HypervisorTemplateAdapter.java:668]: added the check @shwstppr, please review it again.

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/template/HypervisorTemplateAdapter.java [hunk 2]
```diff
@@ -665,8 +669,13 @@ public boolean delete(TemplateProfile profile) {
             // Remove template details
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/template/HypervisorTemplateAdapter.java:669]: letś delete this comment

File: server/src/main/java/com/cloud/template/HypervisorTemplateAdapter.java [hunk 3]
```diff
@@ -665,8 +669,13 @@ public boolean delete(TemplateProfile profile) {
...
             templateDetailsDao.removeDetails(template.getId());
 
-            // Remove deploy-as-is details (if any)
-            templateDeployAsIsDetailsDao.removeDetails(template.getId());
+            // Remove deploy-as-is details (if any and if there are no VMs using it)
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/template/HypervisorTemplateAdapter.java:672]: letś make this a javadoc of a new method

**Review by DaanHoogland — APPROVED** [review]
some style critique but clgtm

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Review by shwstppr — APPROVED** [review]
Code LGTM

**Comment by shwstppr** [comment]
@harikrishna-patnala are you making any more changes for review comments?

File: server/src/main/java/com/cloud/template/HypervisorTemplateAdapter.java (hunk 3 above)
Comment (harikrishna-patnala) [reply server/src/main/java/com/cloud/template/HypervisorTemplateAdapter.java]: Added a new method and java doc

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test matrix

**Comment by DaanHoogland** [comment]
@blueorangutan test rocky8 xenserver-71

**Comment by DaanHoogland** [comment]
@blueorangutan test alma9 vmware-70u3

**Review by rajujith — APPROVED** [review]
LGTM. Instance starts fine with this fix although the template (vAPP) is deleted, it was failing earlier. Also observed is that the data is retained in the table `template_deploy_as_is_details` even after the template deletion as required by this fix. As mentioned already here this has side effects.

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test alma9 vmware-70u3

**Review by weizhouapache — APPROVED** [review]
code lgtm

**Comment by harikrishna-patnala** [comment]
can this PR be merged, I see the above failures on other PRs also

**Comment by DaanHoogland** [comment]
> can this PR be merged, I see the above failures on other PRs also

yes, we shouldn't ignore them, but I am going through them and will add any cleanup or added robustness in #8196

**Pull request merged**